src/pyhf/modifiers/shapesys.py

```diff
@@ -115,30 +116,36 @@ def _precompute(self):
         self.shapesys_default = tensorlib.ones(tensorlib.shape(self.shapesys_mask))
 
     def finalize(self, pdfconfig):
-        for uncert_this_mod, pname in zip(self.__shapesys_uncrt, self._shapesys_mods):
+        # self.__shapesys_info: (parameter, sample, (mask, nominal rate, uncertainty), bin)
+        for mod_uncert_info, pname in zip(self.__shapesys_info, self._shapesys_mods):
+            # a case where given shapesys modifier affects zero samples
             if not pdfconfig.param_set(pname).n_parameters:
                 continue
-            unc_nom = default_backend.astensor(
-                [x for x in uncert_this_mod[:, :, :] if any(x[0][x[0] > 0])]
-            )
-            unc = unc_nom[0, 0]
-            nom = unc_nom[0, 1]
-            unc_sq = default_backend.power(unc, 2)
-            nom_sq = default_backend.power(nom, 2)
-
-            # the below tries to filter cases in which
-            # this modifier is not used by checking non
-            # zeroness.. shoudl probably use mask
-            numerator = default_backend.where(
-                unc_sq > 0, nom_sq, default_backend.zeros(unc_sq.shape)
-            )
-            denominator = default_backend.where(
-                unc_sq > 0, unc_sq, default_backend.ones(unc_sq.shape)
-            )
-
-            factors = numerator / denominator
-            factors = factors[factors > 0]
+
+            # identify the sample that the given parameter affects
+            # shapesys is not shared, so there should only ever be at most one sample
+            sample_uncert_info = mod_uncert_info[
+                default_backend.astensor(
```

Comment:
we already solve a similar problem here `singular_sample_index` maybe we justs save it as an attribute?

Reply:
would allow us to avoid adding the `sum` extension to new dtypes

Reply:
#785 is filed to handle later